apps/near_app/near_pub_key.c

```diff
@@ -0,0 +1,421 @@
+ * @retval true If all the requested xpubs were exported to the host app
+ * @retval false If the export was interrupted by a P0 event or an invalid query
+ * was received from the host app.
+ */
+static bool send_public_keys(near_query_t *query,
+                             const uint8_t pubkey_list[][NEAR_PUB_KEY_SIZE],
+                             const pb_size_t count,
+                             const pb_size_t which_request);
+
+/**
+ * @details The function provides a public key. It accepts NULL for output
+ * parameter and handles accordingly. The function also manages all the terminal
+ * errors during derivation/encoding, in which case it will return false and
+ * send a relevant error to the host closing the request-response pair
+ * TODO: Make this a common utility function
+ *
+ * @param seed Reference to the wallet seed generated
+ * @param path Derivation path of the node to be derived
+ * @param path_length Expected length of the provided derivation path
+ * @param public_key Storage location for raw uncompressed public key
+ *
+ * @retval false If derivation failed
+ */
+static bool get_public_key(const uint8_t *seed,
+                           const uint32_t *path,
+                           uint32_t path_length,
+                           uint8_t *public_key);
+
+/**
+ * @brief Helper function to take user consent before exporting public keys to
+ * the host. Uses an appropriate message template based on the query request
+ * received from the host.
+ *
+ * @param which_request The type of request received from host
+ * @param wallet_name The name of the wallet on which the request needs to be
+ * performed
+ * @return true If the user accepted the request
+ * @return false If the user rejected or any P0 event occurred during the
+ * confirmation.
+ */
+static bool get_user_consent(const pb_size_t which_request,
+                             const char *wallet_name);
+
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+static bool check_which_request(const near_query_t *query,
+                                pb_size_t which_request) {
+  if (which_request != query->get_public_key.which_request) {
+    near_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                    ERROR_DATA_FLOW_INVALID_REQUEST);
+    return false;
+  }
+
+  return true;
+}
+
+static bool validate_request(const near_get_public_key_intiate_request_t *req,
+                             const pb_size_t which_request) {
+  bool status = true;
+  const pb_size_t count = req->derivation_paths_count;
+
+  if (0 == count) {
+    // request does not have any derivation paths, invalid request
+    near_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                    ERROR_DATA_FLOW_INVALID_DATA);
+    status = false;
+  }
+
+  if (NEAR_QUERY_GET_PUBLIC_KEYS_TAG == which_request && 1 < count) {
```

Comment:
Replace with `NEAR_QUERY_GET_PUBLIC_KEY_TAG` instead of `NEAR_QUERY_GET_PUBLIC_KEYS_TAG`

Reply:
Thanks for this catch.